### PR TITLE
Husam Dayya - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by husamdev**

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+//  MVP
+//
+//  Created by Husam Dayya on 14/05/2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+
+struct FeedErrorViewModel {
+	let errorMessage: String
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift:12]: The `errorMessage` could be an optional `String?`, so the Presenter can decide when to show/hide the message.

For example, in the `didStartLoadingFeed()` method, the Presenter can call  `feedErrorView.display(FeedErrorViewModel(errorMessage: .none))` to hide the error.

This way, the ViewController doesn't need to decide when to hide the error. It's easier to manage presentation state transitions when all changes are centralized in the presenter.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -22,6 +24,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	}
 	
 	@IBAction private func refresh() {
+		errorView.hideMessage()
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:27]: The Presenter should decide when to hide/show errors. See my other comment.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 3]
```diff
@@ -8,9 +8,11 @@ protocol FeedViewControllerDelegate {
 	func didRequestFeedRefresh()
 }
 
-public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching, FeedLoadingView {
+public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching, FeedLoadingView, FeedErrorView {
 	var delegate: FeedViewControllerDelegate?
 	
+	@IBOutlet weak var errorView: ErrorView!
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:14]: This property doesn't need to be `weak`. Also, the setter can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView!
```

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 4]
```diff
@@ -20,15 +20,24 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onFeedLoadFailure: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onFeedLoadFailure: Observer<String>?` could be `onErrorStateChange: Observer<String?>?`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's easier to manage presentation state transitions when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -5,6 +5,9 @@
 import UIKit
 
 public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching {
+	
+	@IBOutlet weak var errorView: ErrorView!
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:9]: This property doesn't need to be `weak`. Also, the setter can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView!
```

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 6]
```diff
@@ -21,6 +24,7 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 	
 	@IBAction private func refresh() {
 		viewModel?.loadFeed()
+		errorView.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:27]: The ViewModel should decide when to hide/show errors. See my other comment.

File: MVP Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift [hunk 7]
```diff
@@ -57,4 +57,12 @@ extension FeedViewController {
...
 		return 0
 	}
+	
+	var errorMessage: String? {
+		return (tableView.tableHeaderView as? ErrorView)?.message
+	}
+	
+	func simulateTapOnErrorMessage() {
+		(tableView.tableHeaderView as? ErrorView)?.button.simulateTap()
+	}
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/Helpers/FeedViewController+TestHelpers.swift:67]: You can create a test helper to extract the error view casting logic. The same way you did it in the MVVM solution:

```swift
private var errorView: ErrorView? {
    tableView.tableHeaderView as? ErrorView
}
```

**Review by caiozullo — COMMENTED** [review]
Hi @husamdev, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by husamdev** [comment]
Hey Caio!
Can you please check my PR now?

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 8]
```diff
@@ -33,6 +35,10 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		}
 	}
 	
+	func display(_ viewModel: FeedErrorViewModel) {
+		errorView.show(message: viewModel.errorMessage)
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:39]: Here the view controller needs to check if there's a message or not.

If there's a message, the view controller should call `show()`
If the message is `nil`, the view controller should call `hideMessage()`

```suggestion
		if let message = viewModel.errorMessage {
			errorView.show(message: message)
		} else {
			errorView.hideMessage()
		}
```

File: MVP Module/Source/Feed UI/Views/ErrorView.swift [hunk 9]
```diff
@@ -22,7 +22,7 @@ public final class ErrorView: UIView {
 		alpha = 0
 	}
 	
-	func show(message: String) {
+	func show(message: String?) {
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Views/ErrorView.swift:25]: The `ErrorView.show` method shouldn't be changed (see my previous comment). 

Please, undo the change here.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 10]
```diff
@@ -5,6 +5,11 @@
 import Foundation
 import FeedFeature
 
+enum FeedErrorViewModel {
+	case show(message: String)
+	case hide
+}
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:11]: This `enum` is exactly the same as the `Optional` enum.

So you could use an optional `String?` instead to make the code more concise.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 11]
```diff
@@ -20,15 +25,24 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onFeedLoadFailure: Observer<FeedErrorViewModel>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:28]: As described in my previous comment, `onFeedLoadFailure: Observer<FeedErrorViewModel>?` could be `onErrorStateChange: Observer<String?>?` - noticed the optional `String?` type.

This way, you don't need to create a custom enum. Moreover, the name `onErrorStateChange` is more expressive since it also handles the 'no error' case with `nil`.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 12]
```diff
@@ -20,15 +25,24 @@ final class FeedViewModel {
...
 	
 	func loadFeed() {
 		onLoadingStateChange?(true)
+		onFeedLoadFailure?(.hide)
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:32]: When using the optional type, instead of `.hide` you can use `nil` or `.none`:

```swift
onErrorStateChange?(.none)
```

**Review by caiozullo — COMMENTED** [review]
Hi @husamdev, almost there!

Please, check my latest review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by husamdev** [comment]
Hello @caiozullo 
I believe it's ready now!
Thank you in advance

**Review by caiozullo — COMMENTED** [review]
Hi @husamdev,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

Great job!
